src/lib/replication/repl_dev/raft_state_machine.cpp

```diff
@@ -0,0 +1,510 @@
+        (rreq->value.size ? rreq->local_blkid.serialized_size() : 0);
+    rreq->alloc_journal_entry(entry_size, true /* raft_buf */);
+    rreq->journal_entry->code = (rreq->value.size) ? journal_type_t::HS_LARGE_DATA : journal_type_t::HS_HEADER_ONLY;
+    rreq->journal_entry->user_header_size = rreq->header.size();
+    rreq->journal_entry->key_size = rreq->key.size();
+    rreq->journal_entry->dsn = rreq->dsn();
+
+    uint8_t* raw_ptr = uintptr_cast(rreq->journal_entry) + sizeof(repl_journal_entry);
+    if (rreq->header.size()) {
+        std::memcpy(raw_ptr, rreq->header.cbytes(), rreq->header.size());
+        raw_ptr += rreq->header.size();
+    }
+
+    if (rreq->key.size()) {
+        std::memcpy(raw_ptr, rreq->key.cbytes(), rreq->key.size());
+        raw_ptr += rreq->key.size();
+    }
+
+    if (rreq->value.size) {
+        auto const b = rreq->local_blkid.serialize();
+        std::memcpy(raw_ptr, b.cbytes(), b.size());
+        raw_ptr += b.size();
+    }
+
+    auto* vec = sisl::VectorPool< raft_buf_ptr_t >::alloc();
+    vec->push_back(rreq->raft_journal_buf());
+
+    nuraft::raft_server::req_ext_params param;
+    param.after_precommit_ = bind_this(RaftStateMachine::after_precommit_in_leader, 1);
+    param.expected_term_ = 0;
+    param.context_ = voidptr_cast(rreq.get());
+
+    m_rd.raft_server()->append_entries_ext(*vec, param);
+    sisl::VectorPool< raft_buf_ptr_t >::free(vec);
+}
+
+repl_req_ptr_t RaftStateMachine::transform_journal_entry(nuraft::ptr< nuraft::log_entry >& lentry) {
+    // Leader has nothing to transform or process
+    if (m_rd.is_leader()) { return nullptr; }
+
+    repl_journal_entry* jentry = r_cast< repl_journal_entry* >(lentry->get_buf().data_begin());
+    RELEASE_ASSERT_EQ(jentry->major_version, repl_journal_entry::JOURNAL_ENTRY_MAJOR,
+                      "Mismatched version of journal entry received from RAFT peer");
+
+    // For inline data we don't need to transform anything
+    if (jentry->code != journal_type_t::HS_LARGE_DATA) { return nullptr; }
+
+    sisl::blob const header = sisl::blob{uintptr_cast(jentry) + sizeof(repl_journal_entry), jentry->user_header_size};
+    sisl::blob const key = sisl::blob{header.cbytes() + header.size(), jentry->key_size};
+    DEBUG_ASSERT_GT(jentry->value_size, 0, "Entry marked as large data, but value size is notified as 0");
+
+    // From the repl_key, get the repl_req. In cases where log stream got here first, this method will create a new
+    // repl_req and return that back. Fill up all of the required journal entry inside the repl_req
+    auto rreq = m_rd.follower_create_req(
+        repl_key{.server_id = jentry->server_id, .term = lentry->get_term(), .dsn = jentry->dsn}, header, key,
+        jentry->value_size);
+    rreq->journal_buf = lentry->serialize();
+
+    MultiBlkId entry_blkid;
+    entry_blkid.deserialize(sisl::blob{key.cbytes() + key.size(), jentry->value_size}, true /* copy */);
```

Comment:
jentry->value_size is the data value size, right? actually what we need here is the PBA serialize size.

Reply:
I think this has been take care now